igwn_auth_utils/requests.py

```diff
@@ -0,0 +1,360 @@
+        if auth is not False:  # not disabled
+            self.auth = self._find_username_password(url)
+
+        return self.auth
+
+    def _set_token_header(self, token):
+        """Serialise a `scitokens.SciToken` and format and store as an
+        Authorization header for this session.
+
+        Parameters
+        ----------
+        token : `scitokens.SciToken`, `str`, `bytes`
+            the token to serialize, or an already serialized representation
+        """
+        if isinstance(token, (str, bytes)):  # load a valid token
+            header = f"Bearer {token}"
+        else:
+            header = scitoken_authorization_header(token)
+        self.headers["Authorization"] = header
+        return header
+
+    @staticmethod
+    def _find_x509_credentials(error=True):
+        """Find an X.509 certificate for authorization
+        """
+        return _find_cred(find_x509_credentials, error=error)
+
+    @staticmethod
+    def _find_token(audience, scope, error=True):
+        """Find a bearer token for authorization
+        """
+        return _find_cred(find_scitoken, audience, scope, error=error)
+
+    @staticmethod
+    def _find_username_password(url):
+        """Use `netrc.netrc` to find the username/password for basic auth
+        """
+        host = urlparse(url).hostname
+
+        try:
+            result = netrc(getenv("NETRC")).authenticators(host)
+        except IOError:
+            return
+
+        try:
+            username, _, password = result
+        except TypeError:
+            return
+        return username, password
+
+
+class Session(
+    SessionAuthMixin,
+    SessionErrorMixin,
+    requests.Session,
+):
+    """`requests.Session` class with default IGWN authorization handling
+
+    {parameters}
+
+    Examples
+    --------
+    >>> from igwn_auth_utils.requests import Session
+    >>> with Session(token=mytoken) as sess:
+    ...     sess.get("https://my.science.stuff/data")
+    """
+
+
+# update the docstrings to include the same parameter info
+for _obj in (Session, SessionAuthMixin):
+    _obj.__doc__ = _obj.__doc__.format(parameters=_auth_session_parameters)
+
+
+def get(url, *args, session=None, **kwargs):
+    """Request data from a URL via an HTTP ``'GET'`` request
+
+    Parameters
+    ----------
+    url : `str`,
+        the URL to request
+
+    session : `requests.Session`, optional
+        the connection session to use, if not given one will be
+        created on-the-fly
+
+    *args, **kwargs
+        all other keyword arguments are passed directly to
+        `requests.Session.get`
+
+    Returns
+    -------
+    resp : `requets.Response`
+        the response object
+
+    See also
+    --------
+    requests.Session.get
+        for information on how the request is performed
+    """
+    # user's session
+    if session:
+        return session.get(url, *args, **kwargs)
+
+    # new session
+    sess_kwargs = {k: kwargs.pop(k, None) for k in (
```

Comment:
Don't modify kwargs. Or if you do, make a copy of it. The calling code probably does not expect this function to have side effects of modifying kwargs.

Reply:
`kwargs` is internal to the scope of this function and so cannot impact the calling code.

Reply:
Oh, you're right! I learn something new every day...

```python
foo = {'bar': 1, 'bat': 2}

def xyzzy(**kwargs):
    kwargs.pop('bat')

xyzzy(**foo)
print(foo)  # outputs {'bar': 1, 'bat': 2}
```